### PR TITLE
Add ability to auto-generate examples from tests

### PR DESCRIPTION
**Pull request by m-hurlow**
This PR adds the ability to insert usage examples into the user docs which are auto-generated from a test file.

All the new code is in `loader_fetch.py`. Before `_fetch_and_process_file()` is called, `_fetch_test_file()` attempts to find a file called `eval_tests.yaml` in the repository root. If this file exists, it is parsed and stored. Later, in `_edit_user_docs()`, if the test file was found, the tests inside are processed and a new section is added to the document with examples generated from the tests. 

This process is decoupled from the actual testing, aside from using the same file, so any testing framework can still be used.

There is a new dependency on a library called "autotests", which provides a class that parses the YAML file and abstracts its contents. This is intended to allow the format to be altered without breaking compatibility with every eval function that uses it. Currently this is stored in a [repo](https://github.com/m-hurlow/autotests/) owned by me, but I would like to move it into the lambda-feedback organisation soon.

### CODE REVIEW

**Comment by aplr** [comment]
Great! will do a more thorough review. Some initial thoughts:

1. Do you think it would also make sense to put autotest into the toolkit repo, to avoid maintaining lots of packages?
2. Could you create a [json-schema](https://json-schema-everywhere.github.io/yaml) for the eval_tests.yaml, so you get hints in your IDE? In each of the template repos, we can add a `.vscode/settings.json`, which [configures the json schema](https://code.visualstudio.com/docs/languages/json#_mapping-in-the-user-settings) to use for the `eval_tests.yaml`. We just need to figure out where to place the spec, as both `autotest` and `python-toolkit` is python-specific, but the `eval_tests.yaml` should be language-independent.

File: evaldocsloader/loader_fetch.py [hunk 1]
```diff
@@ -182,6 +186,57 @@ def _edit_user_docs(self, doc: mistletoe.Document) -> mistletoe.Document:
...
+
+                        if sub_test.desc:
+                            doc.children.append(mistletoe.block_token.Paragraph([sub_test.desc]))
+                            doc.children.append(mistletoe.markdown_renderer.BlankLine({}))
+                    
+                    # Flush any remaining examples
+                    if len(table_lines) != 0:
+                        doc.children.append(mistletoe.block_token.Table((table_header + table_lines, 0)))
+                        doc.children.append(mistletoe.markdown_renderer.BlankLine({}))
+
```
Comment (aplr) [inline evaldocsloader/loader_fetch.py]: Maybe move this into another function, e.g. `_edit_user_docs_insert_autotests` so it's easily visible which edits are done by the `_edit_user_docs` as it grows and does more automated edits

File: evaldocsloader/loader_fetch.py [hunk 2]
```diff
@@ -182,6 +186,57 @@ def _edit_user_docs(self, doc: mistletoe.Document) -> mistletoe.Document:
...
+                    # A buffer that allows tables for multiple sub-tests to be combined
+                    table_lines = []
+                    
+                    # Sub tests have the same answer and parameters as a test, but a different response value
+                    for sub_test in test.sub_tests:
+                        if sub_test.exclude_from_docs:
+                            continue
+                        
+                        response = sanitise_response(sub_test.response)
+                        answer = sanitise_response(test.answer)
```
Comment (aplr) [inline evaldocsloader/loader_fetch.py]: Is sanitation necessary, as both `response` and `answer` are anyways wrapped in a inline-code "`"? I tested it with a online markdown editor, and it worked. Or does Mistletoe require this when inserting raw strings?

File: evaldocsloader/loader_fetch.py (hunk 1 above)
Comment (m-hurlow) [reply evaldocsloader/loader_fetch.py]: That's a good point, I've moved it to its own function now.

File: evaldocsloader/loader_fetch.py (hunk 2 above)
Comment (m-hurlow) [reply evaldocsloader/loader_fetch.py]: MkDocs renders `|` characters between backticks correctly, but mistletoe still interprets them as column separators and adds unnecessary columns.
Here's what the raw markdown looks like;
With sanitisation:
```markdown
| Response                   | Answer  | Correct? |
| -------------------------- | ------- | -------- |
| `~(~(~A | ~B) | ~(A | B))` | `A ^ B` | ✓        |
```
Without sanitisation:
```markdown
| Response | Answer | Correct? |      |         |     |
| -------- | ------ | -------- | ---- | ------- | --- |
| `~(~(~A  | ~B)    | ~(A      | B))` | `A ^ B` | ✓   |
```

This adds unnecessary whitespace to the right:

Screenshot 2024-09-06 110315

When it looks like this with sanitisation:

Screenshot 2024-09-06 110442

**Comment by m-hurlow** [comment]
> Great! will do a more thorough review. Some initial thoughts:
> 1. Do you think it would also make sense to put autotest into the toolkit repo, to avoid maintaining lots of packages?
> 2. Could you create a [json-schema](https://json-schema-everywhere.github.io/yaml) for the eval_tests.yaml, so you get hints in your IDE? In each of the template repos, we can add a `.vscode/settings.json`, which [configures the json schema](https://code.visualstudio.com/docs/languages/json#_mapping-in-the-user-settings) to use for the `eval_tests.yaml`. We just need to figure out where to place the spec, as both `autotest` and `python-toolkit` is python-specific, but the `eval_tests.yaml` should be language-independent.

In response to your initial thoughts:

1. I initially thought this, but the libraries appear to have differrent purposes: lf-toolkit is code for writing the functional parts of eval functions, but autotests is purely for testing, and is intended to not only be used by eval functions but also by EvalDocsLoader and any other Python tools that need to read the test YAML files. I didn't think that it would be best for EvalDocsLoader to depend on all the other code in lf-toolkit that it wouldn't use. If you think it would make sense to have it all together though, we can move it over.
2. That's definitely a good idea, and it would help developers to hand-write tests more easily. I will work on doing that.

File: evaldocsloader/loader_fetch.py (hunk 2 above)
Comment (aplr) [reply evaldocsloader/loader_fetch.py]: Hilarious, ok perfectly fine wanted to understand!

**Comment by aplr** [comment]
> 1. I initially thought this, but the libraries appear to have differrent purposes: lf-toolkit is code for writing the functional parts of eval functions, but autotests is purely for testing, and is intended to not only be used by eval functions but also by EvalDocsLoader and any other Python tools that need to read the test YAML files. I didn't think that it would be best for EvalDocsLoader to depend on all the other code in lf-toolkit that it wouldn't use. If you think it would make sense to have it all together though, we can move it over.
> 2. That's definitely a good idea, and it would help developers to hand-write tests more easily. I will work on doing that.

Fine with that!

**Review by aplr — APPROVED** [review]
LGTM

**Pull request merged**